src/test/java/net/neoforged/moddevgradle/tasks/JarJarTest.java

```diff
@@ -34,6 +37,103 @@ public void testNoSourceWhenNoDependenciesAreDefined() throws IOException {
+        assertThat(result.getOutput()).contains("Embedding dependency curse.maven:jade-324717:5444008 from cursemaven using JiJ is likely to cause conflicts at runtime when other mods include the same library from a normal Maven repository.");
+    }
+
+    @Test
+    void testCannotEmbedLocalFileWithoutExplicitJavaModuleName() throws IOException {
+        var localFile = tempDir.resolve("file.jar");
+        new JarOutputStream(Files.newOutputStream(localFile), new Manifest()).close();
+
+        var e = assertThrows(UnexpectedBuildFailure.class, () -> runWithSource("""
+                dependencies {
+                    jarJar(files("file.jar"))
+                }
+                """));
+        assertThat(e).hasMessageFindingMatch("Cannot embed local file dependency .*file.jar because it has no explicit Java module name.\\s*" +
+                                             "Please set either 'Automatic-Module-Name' in the Jar manifest, or make it an explicit Java module.\\s*" +
+                                             "This ensures that your file does not conflict with another mods library that has the same or a similar filename.");
+    }
+
+    @Test
+    void testCanEmbedLocalFileWithAutomaticModuleName() throws Exception {
+        var localFile = tempDir.resolve("file.jar");
+        var manifest = new Manifest();
+        manifest.getMainAttributes().putValue("Manifest-Version", "1.0");
+        manifest.getMainAttributes().putValue("Automatic-Module-Name", "super_duper_module");
+        new JarOutputStream(Files.newOutputStream(localFile), manifest).close();
+        var md5Hash = FileUtils.hashFile(localFile.toFile(), "MD5");
+
+        var result = runWithSource("""
+                dependencies {
+                    jarJar(files("file.jar"))
+                }
+                """);
+        assertEquals(SUCCESS, result.task(":jarJar").getOutcome());
+        assertEquals(new Metadata(
+                List.of(
+                        new ContainedJarMetadata(
+                                new ContainedJarIdentifier("", "super_duper_module"),
+                                new ContainedVersion(VersionRange.createFromVersionSpec("[" + md5Hash + "]"), new DefaultArtifactVersion(md5Hash)),
+                                "META-INF/jarjar/file.jar",
+                                false
+                        )
+                )
+        ), readMetadata());
+    }
+
+    @Test
+    void testCanEmbedLocalFileWithModuleInfo() throws Exception {
+        var moduleInfoJava = tempDir.resolve("src/service/java/module-info.java");
+        Files.createDirectories(moduleInfoJava.getParent());
+        Files.writeString(moduleInfoJava, "module super_duper_module {}");
+
+        var result = runWithSource("""
+                sourceSets {
+                    service
+                }
+                compileServiceJava {
+                    // otherwise testkit needs to run with J21
```

Comment:
Why Java 9 specifically and not 17 which MDG uses? Regardless, you could use a toolchain for Java 21.